### PR TITLE
CLOUDP-36241 Conf call integration

### PR DESCRIPTION
**Pull request by chunmingli**
https://jira.mongodb.org/browse/CLOUDP-36241

Introduced a `confCallMode` flag to the manual mode. If `file`, conf call _response_ is served from the conf.yaml file. If `remote`, the PIT module will still read from conf.yaml, but only for the conf call params bit, and then the initial conf call _response_ is fetched from remote MMS.

For I also provided default values in the PIT module where the fields of the initial conf call response is missing from file/remote. But when a field exists in the conf call response, it takes priority over the default value, effectively overrides the default value.

Can run with:
```
./cps-pit -config conf.yaml -confCallMode [remote|file]

defaulting to file mode

./cps-pit -config conf.yaml
```

### CODE REVIEW

File: cps-pit/op_reader.go [hunk 1]
```diff
@@ -44,11 +51,5 @@ func runReader(
 		}
 		println(raw.String())
 	}
-}
-
-func getDoc(ctx context.Context, cur mongo.Cursor) (bson.Raw, error) {
-	if cur.Next(ctx) {
-		return cur.DecodeBytes()
-	}
-	return nil, nil
```
Comment (chunmingli) [inline cps-pit/op_reader.go]: Drive-by fix, should not have returned nil doc in this case.

File: cps-pit/op_reader.go [hunk 2]
```diff
@@ -29,13 +30,19 @@ func runReader(
...
-		raw, err := getDoc(ctx, cur)
+	for cur.Next(ctx) {
+		raw, err := cur.DecodeBytes()
 		if err != nil {
 			println("error in produceDocs", err)
 			errc <- err
 			return
 		}
+
+		if raw == nil {
```
Comment (chunmingli) [inline cps-pit/op_reader.go]: Paranoia nil check

File: cps-pit/pusher.go [hunk 3]
```diff
@@ -73,6 +74,12 @@ func (pusher *Pusher) Run(
 					return
 				}
 
+				if slice == nil {
```
Comment (chunmingli) [inline cps-pit/pusher.go]: Paranoia nil check

File: cps-pit/conf.yaml [hunk 4]
```diff
@@ -1,6 +1,28 @@
-bucket:
-region:
-accessKeyId:
-secretAccessKey:
-sessionToken:
-expiration:
+confCallParams:
```
Comment (chunmingli) [inline cps-pit/conf.yaml]: Provided more logical grouping in the config file.

File: cps-pit/conf.yaml [hunk 5]
```diff
@@ -1,6 +1,28 @@
...
+    confCallProtocolScheme: http
+    confCallHostname: localhost
+    confCallPort: 8080
+
+    agentKey:
+    groupId:
+    clusterName: Cluster0
+    rsId: 0
+
+sourceHostname: localhost
```
Comment (lawctan) [inline cps-pit/conf.yaml]: I think we should put the conf call response fields into one group as well. That way, we can easily tell which fields can override conf call

File: cps-pit/config.go [hunk 6]
```diff
@@ -7,40 +7,132 @@ import (
...
+
+	return &(wrapper.ConfCallRequestParams), nil
+}
+
+func ParseConfigFile(filename string, logger *slogger.Logger) (*PitConfig, error) {
+	handle, err := os.Open(filename)
+	if err != nil {
+		return nil, logger.Errorf(slogger.ERROR, "Error opening the config file. Filename: %v Err: %v", filename, err)
+	}
+
```
Comment (lawctan) [inline cps-pit/config.go]: `defer handle.close()`

File: cps-pit/cps-pit.go [hunk 7]
```diff
@@ -14,23 +15,52 @@ func getLogger() *slogger.Logger {
...
+	if *mode == "remote" {
+		var confCallParams *ConfCallRequestParams
+		confCallParams, err = ParseConfCallParams(*configFilename, logger)
+		if err != nil {
+			logger.Logf(slogger.ERROR, "Error parsing all required settings: %v\n", err)
+			os.Exit(1)
+		}
+		println(confCallParams.String())
+
+		pitConfig, err = RemoteConfCall(confCallParams, logger)
```
Comment (lawctan) [inline cps-pit/cps-pit.go]: where is this defined?

File: cps-pit/mms_client.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+				conn.SetDeadline(time.Now().Add(connDeadline))
+				return conn, err
+			},
+			MaxIdleConns:          maxIdleConns,
+			DisableKeepAlives:     !keepAlive,
+			ResponseHeaderTimeout: responseHeaderTimeout,
+		},
+	}
+
+	return client
```
Comment (lawctan) [inline cps-pit/mms_client.go]: not making this a global variable? don't see it being used either

File: cps-pit/cps-pit.go (hunk 7 above)
Comment (lawctan) [reply cps-pit/cps-pit.go]: i see it now. will take another look

File: cps-pit/mms_client.go (hunk 8 above)
Comment (lawctan) [reply cps-pit/mms_client.go]: i see it now. will take another look

File: cps-pit/conf_call.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/mongodb/slogger/v2/slogger"
+)
+
+func (p *ConfCallRequestParams) ConfUrl() string {
```
Comment (lawctan) [inline cps-pit/conf_call.go]: Two differences I noticed compared to backup agent:
1) Splitting up mothership into different files
2) Creating a client for every caller instead of having one global httpClient. I guess this is a good thing because we're avoiding global variables.

I don't have strong opinion on using either approach. But just curious about your thoughts on why you made those changes.

File: cps-pit/conf.yaml (hunk 5 above)
Comment (chunmingli) [reply cps-pit/conf.yaml]: I went back and forth with whether to split conf call request params into a separate conf file, that would require a bunch of boiler plate code for parsing a new flag and error checking. Hence reusing the single config file for now. So this effectively makes this conf.yaml dual-purpose.

Grouping introduces boiler plate code too due to the top level tag parsing - probably fewer than the split conf files approach.

File: cps-pit/conf_call.go (hunk 9 above)
Comment (chunmingli) [reply cps-pit/conf_call.go]: I don't want to use the same client for everything, and the global variable reason too. I'd to have some flexibility that client can be created on demand and we can later decide the life cycle for each http client instance.

**Comment by chunmingli** [comment]
Comments addressed. @lawctan

**Comment by chunmingli** [comment]
Ping for review. Some of these can potentially be used in [CLOUDP-36248](https://jira.mongodb.org/browse/CLOUDP-36248)
